### PR TITLE
[DEV-3177] email verification

### PR DESCRIPTION
**Pull request by lapatric**
#### Pre-Release

- [ ] Check migrations
  - No database related infos (sqldb-xxx)
  - Impact on GS (new/removed columns)
- [ ] Check for linter errors (in PR)
- [ ] Test basic user operations (on [DFX services](https://dev.services.dfx.swiss))
  - Login/logout
  - Buy/sell payment request
  - KYC page

#### Post-Release

- Test basic user operations
- Monitor application insights log

### CODE REVIEW

File: src/subdomains/generic/user/models/user/user.service.ts [hunk 1]
```diff
@@ -210,11 +225,54 @@ export class UserService {
     });
     if (!userData) throw new NotFoundException('User not found');
 
+    if (dto.mail) {
+      await this.tfaService.checkVerification(userData, ip, TfaLevel.BASIC);
+
+      const user = await this.userRepo.findOneBy({ userData: { mail: dto.mail } });
+      if (user) throw new BadRequestException('Mail already in use');
```
Comment (Yannick1712) [inline src/subdomains/generic/user/models/user/user.service.ts:232]: Should we maybe check on this point if a merge for the userDatas would be possible? 🤔 
We have a entity method in userData `checkIfMergePossibleWith` we could use for this

File: src/subdomains/generic/user/models/user/user.service.ts [hunk 2]
```diff
@@ -564,4 +622,38 @@ export class UserService {
...
+              { key: MailKey.SPACE, params: { value: '4' } },
+              { key: MailKey.DFX_TEAM_CLOSING },
+            ],
+          },
+        });
+    } catch (e) {
+      this.logger.error(`Failed to send verification mail ${userData.id}:`, e);
+      throw new BadRequestException('Failed to send verification mail');
+    }
+  }
```
Comment (Yannick1712) [inline src/subdomains/generic/user/models/user/user.service.ts:658]: I would remove the method here. It's nearly the same as `sendVerificationMail` in tfaService. I would use the method in tfaService, make it public and add a parameter to differentiate between EmailVerification and defaultVerification

File: src/subdomains/generic/user/models/user/user.service.ts [hunk 3]
```diff
@@ -223,7 +286,10 @@ export class UserService {
   }
 
   async updateUserData(id: number, dto: KycInputDataDto): Promise<{ user: UserDetailDto; isKnownUser: boolean }> {
-    const user = await this.userRepo.findOne({ where: { id }, relations: ['userData', 'userData.users', 'wallet'] });
+    const user = await this.userRepo.findOne({
+      where: { id },
+      relations: ['userData', 'userData.users', 'wallet'],
```
Comment (Yannick1712) [inline src/subdomains/generic/user/models/user/user.service.ts:291]: `{userData: {users: {true}}, wallet: {true}}`

File: src/subdomains/generic/kyc/controllers/kyc.controller.ts [hunk 4]
```diff
@@ -335,6 +341,13 @@ export class KycController {
     return this.tfaService.verify(code, dto.token, ip);
   }
 
+  @Get('2fa')
+  @ApiCreatedResponse({ description: '2FA active' })
+  @UseGuards(AuthGuard(), new RoleGuard(UserRole.ACCOUNT))
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/controllers/kyc.controller.ts:346]: `@ApiOkResponse`

File: src/subdomains/generic/user/user.module.ts [hunk 5]
```diff
@@ -76,6 +77,7 @@ import { WebhookService } from './services/webhook/webhook.service';
     BankDataRepository,
     UserDataRelationRepository,
     AccountMergeRepository,
+    TfaLogRepository,
     UserService,
```
Comment (davidleomay) [inline src/subdomains/generic/user/user.module.ts:81]: Remove `TfaLogRepository`

File: src/subdomains/generic/kyc/services/tfa.service.ts [hunk 6]
```diff
@@ -195,6 +188,30 @@ export class TfaService {
         });
     } catch (e) {
       this.logger.error(`Failed to send verification mail ${userData.id}:`, e);
+      throw new BadRequestException('Failed to send verification mail');
     }
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/tfa.service.ts:192]: `ServiceNotAvailableException`?

File: src/subdomains/generic/user/models/user-data/user-data.service.ts [hunk 7]
```diff
@@ -285,16 +313,46 @@ export class UserDataService {
...
+    this.secretCache.delete(userData.id);
+
+    return this.updateUserMailInternal(userData, { mail: cacheEntry.mail });
+  }
+
   async updateUserSettings(
     userData: UserData,
     dto: UpdateUserDto,
     forceUpdate?: boolean,
   ): Promise<{ user: UserData; isKnownUser: boolean }> {
```
Comment (davidleomay) [inline src/subdomains/generic/user/models/user-data/user-data.service.ts:353]: `forceUpdate` parameter can be removed

File: src/subdomains/generic/user/models/user-data/user-data.service.ts [hunk 8]
```diff
@@ -309,26 +367,22 @@ export class UserDataService {
...
       }
     }
 
-    if (mailChanged) await this.kycLogService.createMailChangeLog(userData, userData.mail, dto.mail);
-
     userData = await this.userDataRepo.save(Object.assign(userData, dto));
 
-    const isKnownUser = (mailChanged || forceUpdate) && (await this.isKnownKycUser(userData));
+    const isKnownUser = forceUpdate && (await this.isKnownKycUser(userData));
     return { user: userData, isKnownUser };
```
Comment (davidleomay) [inline src/subdomains/generic/user/models/user-data/user-data.service.ts:386]: The `isKnownUser` check can be removed here since the mail doesn't change

**Review by davidleomay — APPROVED** [review]

**Pull request merged**